[Android] Add `.cxx` directories to app template `.gitignore`

gmackall opened the pull request:
Fixes (but I still need to write the draft of the blog section so don't autoclose github 🙏 ) https://github.com/flutter/flutter/issues/160372.

-Module templates [already exclude the entire `.android` directory](https://github.com/flutter/flutter/blob/d9b7e5664604a55a3e900b0c619ee319065abb0e/packages/flutter_tools/templates/module/common/.gitignore.tmpl#L40).
-Plugin templates [already exclude `.cxx`](https://github.com/flutter/flutter/blob/d9b7e5664604a55a3e900b0c619ee319065abb0e/packages/flutter_tools/templates/plugin_shared/android.tmpl/.gitignore).

Modified the existing gitignores with
```
nl=$'\n'
sed -i'' -e '/GeneratedPluginRegistrant.java/a\'$'\n''.cxx/'"\\${nl}"'' $(find . -path \*android/.gitignore)
```
(which was probably not the best way)

Fun fact: apparently the `.gitignore` files (including the ones we copy to the template apps) aren't indexed for search (the `.gitignore.tmpl` are, but not the ones that we copy exactly [such as this one](https://github.com/flutter/flutter/blob/27ba2f279075a9a66b48ce78d94d2999b06d6e26/packages/flutter_tools/templates/app/android.tmpl/.gitignore)).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

mackall-work committed:
add cxx to app template

gmackall committed:
Merge branch 'master' into gitignore_cxx_dirs

mackall-work committed:
redo sed replacement in existing gitignores

gnprice commented on packages/flutter_tools/templates/app/android.tmpl/.gitignore:
```diff
@@ -5,6 +5,7 @@ gradle-wrapper.jar
 /gradlew.bat
 /local.properties
 GeneratedPluginRegistrant.java
+.cxx
```
nit:
```suggestion
.cxx/
```

It feels a bit cleaner to be explicit that this is a directory. (And I see the versions added to the test and example apps already have that.)

gnprice reviewed (commented):
Thanks!

One nit below. I also did a quick search for existing gitignore files this should perhaps be added to:
`$ git ls-files | grep android/.gitignore`
Other than the ones already in this PR, that finds:
```
engine/src/flutter/shell/platform/android/.gitignore
engine/src/flutter/testing/scenario_app/android/.gitignore
packages/integration_test/android/.gitignore
```

I think the second one is a true match that should get this change. The third I'm not sure about. The first is a false positive.


Oh and this search for relevant templates:
```
$ git ls-files packages/flutter_tools/ | grep android.*gitignore
packages/flutter_tools/templates/app/android.tmpl/.gitignore
packages/flutter_tools/templates/plugin/android.tmpl/.gitignore
packages/flutter_tools/templates/plugin_shared/android.tmpl/.gitignore
```

The `app/` one is included here, and `plugin_shared/` already has this line. I don't understand the difference between `plugin/` and `plugin_shared/` — should `plugin/` get this too?

reidbaker committed:
Update packages/flutter_tools/templates/app/android.tmpl/.gitignore

Co-authored-by: Greg Price <gnprice@gmail.com>

mackall-work committed:
two more

gmackall commented:
> Thanks!
> 
> One nit below. I also did a quick search for existing gitignore files this should perhaps be added to: `$ git ls-files | grep android/.gitignore` Other than the ones already in this PR, that finds:
> 
> ```
> engine/src/flutter/shell/platform/android/.gitignore
> engine/src/flutter/testing/scenario_app/android/.gitignore
> packages/integration_test/android/.gitignore
> ```
> 
> I think the second one is a true match that should get this change. The third I'm not sure about. The first is a false positive.
> 
> Oh and this search for relevant templates:
> 
> ```
> $ git ls-files packages/flutter_tools/ | grep android.*gitignore
> packages/flutter_tools/templates/app/android.tmpl/.gitignore
> packages/flutter_tools/templates/plugin/android.tmpl/.gitignore
> packages/flutter_tools/templates/plugin_shared/android.tmpl/.gitignore
> ```
> 

`scenario_app` actually doesn't use the flutter tool, so I don't think it should be impacted here by the change to force ndk download (it's days are [numbered regardless](https://github.com/flutter/flutter/pull/160992) - it was a way to test behavior in something like a full flutter app, and essentially replicated whatever framework/tool functionality was needed for the test). 

`packages/integration_test/android/.gitignore` - I added it here as well. I don't know that it will matter (I haven't seen the cxx directories being created by my change for plugin directories), but our plugin templates include it in the gitignore so I think it is fine to add it anyways.

> The `app/` one is included here, and `plugin_shared/` already has this line. I don't understand the difference between `plugin/` and `plugin_shared/` — should `plugin/` get this too?

Also added to the `plugin/` template. My understanding is that generally when both a `foo` and `foo_shared` directory exist, that the output of the template is a function of both of the directories. But I also know that the files that don't end in `tmpl` get copied as is, so I'm not sure how that is possible here (perhaps one of the `.gitignore` files just wins?) Either way, I've checked the output of the template manually and verified that the `.gitignore` contains the `.cxx/` line we want (both before and after adding to the `plugin/` file).

gnprice reviewed (approved):
Sounds good. Thanks!
